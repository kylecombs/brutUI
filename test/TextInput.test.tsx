import * as React from "react";
import {
  render,
  cleanup,
  fireEvent,
  RenderResult,
} from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect";
import { TextInput, TextInputProps } from "../src";
import { FiHelpCircle, FiMail } from "react-icons/fi";

afterEach(cleanup);

interface ISetupTextInput extends RenderResult {
  input: HTMLInputElement;
}

const setup = ({
  handleChange = () => {},
  disabled,
  label,
  type = "email",
  error,
  helperText,
  leadingText,
  LeadingIcon,
  TrailingIcon,
}: Partial<TextInputProps>): ISetupTextInput => {
  const utils = render(
    <TextInput
      type={type}
      value=""
      leadingText={leadingText}
      handleChange={handleChange}
      label={label}
      error={error}
      placeholder="veronica@example.com"
      helperText={helperText}
      LeadingIcon={LeadingIcon}
      TrailingIcon={TrailingIcon}
      disabled={disabled}
    />,
  );

  const input = utils.getByLabelText("input") as HTMLInputElement;

  return {
    input,
    ...utils,
  };
};

describe("TextInput", () => {
  it("Renders TextInput correctly", () => {
    const { asFragment } = setup({
      label: "Email",
      error: "This is an error message",
    });

    expect(asFragment()).toMatchSnapshot();
  });

  it("Renders TextInput with leadingText correctly", () => {
    const { asFragment } = setup({
      type: "text",
      helperText: "This is a hint text to help the user.",
      leadingText: "https://",
    });

    expect(asFragment()).toMatchSnapshot();
  });

  it("Renders TextInput with LeadingIcon and TrailingIcon", () => {
    const { asFragment } = setup({
      LeadingIcon: <FiMail />,
      TrailingIcon: <FiHelpCircle />,
    });

    expect(asFragment()).toMatchSnapshot();
  });

  it("Renders TextInput with only a TrailingIcon", () => {
    const { asFragment } = setup({
      type: "email",
      TrailingIcon: <FiHelpCircle />,
    });

    expect(asFragment()).toMatchSnapshot();
  });

  it("Renders disabled TextInput correctly", () => {
    const { asFragment } = setup({ disabled: true });

    expect(asFragment()).toMatchSnapshot();
  });

  it("Filling in TextInput works", () => {
    const handleChange = jest.fn();

    const { input } = setup({ handleChange });
    const value = "veronica@example.com";

    fireEvent.change(input, { target: { value } });

    expect(handleChange).toHaveBeenCalledWith(value);
  });

  it("Filling in TextInput does not update if disabled", () => {
    const handleChange = jest.fn();

    const { input } = setup({ handleChange, disabled: true });
    const value = "veronica@example.com";

    fireEvent.change(input, { target: { value } });

    expect(handleChange).toHaveBeenCalledTimes(0);
  });
});
