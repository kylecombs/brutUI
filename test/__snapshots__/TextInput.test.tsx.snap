// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextInput Renders TextInput correctly 1`] = `
<DocumentFragment>
  <label
    class="text-sm font-medium mb-1.5 text-gray-700 dark:text-white"
  >
    Email
  </label>
  <div
    class="relative"
  >
    <div
      class="h-11 shadow-sm rounded-lg w-full absolute flex items-center justify-between px-3.5 pointer-events-none"
    >
      <div />
    </div>
    <input
      aria-label="input"
      class="w-full select-none text-gray-900 dark:text-white text-md border h-11 rounded-lg border-error-300 focus:ring-4 focus:border-error-300 focus:ring-error-100 bg-white dark:bg-gray-800"
      placeholder="veronica@example.com"
      type="email"
      value=""
    />
  </div>
  <div
    class="mt-1.5 text-error-500 text-sm"
  >
    This is an error message
  </div>
</DocumentFragment>
`;

exports[`TextInput Renders TextInput with LeadingIcon and TrailingIcon 1`] = `
<DocumentFragment>
  <div
    class="relative"
  >
    <div
      class="h-11 shadow-sm rounded-lg w-full absolute flex items-center justify-between px-3.5 pointer-events-none"
    >
      <svg
        class="text-gray-500"
        fill="none"
        height="1em"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M4 4h16c1.1 0 2 .9 2 2v12c0 1.1-.9 2-2 2H4c-1.1 0-2-.9-2-2V6c0-1.1.9-2 2-2z"
        />
        <polyline
          points="22,6 12,13 2,6"
        />
      </svg>
      <svg
        class="text-gray-400"
        fill="none"
        height="1em"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <circle
          cx="12"
          cy="12"
          r="10"
        />
        <path
          d="M9.09 9a3 3 0 0 1 5.83 1c0 2-3 3-3 3"
        />
        <line
          x1="12"
          x2="12.01"
          y1="17"
          y2="17"
        />
      </svg>
    </div>
    <input
      aria-label="input"
      class="w-full select-none text-gray-900 dark:text-white text-md border h-11 pl-9 pr-9 rounded-lg border-gray-300 dark:border-gray-500 focus:ring-4 focus:border-primary-300 dark:focus:border-gray-100 focus:ring-primary-100 dark:focus:ring-gray-100 dark:focus:ring-opacity-20 bg-white dark:bg-gray-800"
      placeholder="veronica@example.com"
      type="email"
      value=""
    />
  </div>
</DocumentFragment>
`;

exports[`TextInput Renders TextInput with leadingText correctly 1`] = `
<DocumentFragment>
  <div
    class="relative flex items-center"
  >
    <div
      class="h-11 shadow-sm rounded-lg w-full absolute flex items-center justify-between px-3.5 pointer-events-none"
    >
      <div />
    </div>
    <div
      class="flex items-center h-11 text-lg text-gray-500 pl-3.5 pr-3 border border-r-0 rounded-l-lg border-gray-300 dark:border-gray-500 dark:bg-gray-800"
    >
      https://
    </div>
    <input
      aria-label="input"
      class="w-full select-none text-gray-900 dark:text-white text-md border h-11 rounded-l-0 rounded-r-lg border-gray-300 dark:border-gray-500 focus:ring-4 focus:border-primary-300 dark:focus:border-gray-100 focus:ring-primary-100 dark:focus:ring-gray-100 dark:focus:ring-opacity-20 bg-white dark:bg-gray-800"
      placeholder="veronica@example.com"
      type="text"
      value=""
    />
  </div>
  <div
    class="mt-1.5 text-gray-500 text-sm"
  >
    This is a hint text to help the user.
  </div>
</DocumentFragment>
`;

exports[`TextInput Renders TextInput with only a TrailingIcon 1`] = `
<DocumentFragment>
  <div
    class="relative"
  >
    <div
      class="h-11 shadow-sm rounded-lg w-full absolute flex items-center justify-between px-3.5 pointer-events-none"
    >
      <div />
      <svg
        class="text-gray-400"
        fill="none"
        height="1em"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <circle
          cx="12"
          cy="12"
          r="10"
        />
        <path
          d="M9.09 9a3 3 0 0 1 5.83 1c0 2-3 3-3 3"
        />
        <line
          x1="12"
          x2="12.01"
          y1="17"
          y2="17"
        />
      </svg>
    </div>
    <input
      aria-label="input"
      class="w-full select-none text-gray-900 dark:text-white text-md border h-11 pr-9 rounded-lg border-gray-300 dark:border-gray-500 focus:ring-4 focus:border-primary-300 dark:focus:border-gray-100 focus:ring-primary-100 dark:focus:ring-gray-100 dark:focus:ring-opacity-20 bg-white dark:bg-gray-800"
      placeholder="veronica@example.com"
      type="email"
      value=""
    />
  </div>
</DocumentFragment>
`;

exports[`TextInput Renders disabled TextInput correctly 1`] = `
<DocumentFragment>
  <div
    class="relative"
  >
    <div
      class="h-11 shadow-sm rounded-lg w-full absolute flex items-center justify-between px-3.5 pointer-events-none"
    >
      <div />
    </div>
    <input
      aria-label="input"
      class="w-full select-none text-gray-900 dark:text-white text-md border h-11 rounded-lg border-gray-300 dark:border-gray-500 focus:ring-4 focus:border-primary-300 dark:focus:border-gray-100 focus:ring-primary-100 dark:focus:ring-gray-100 dark:focus:ring-opacity-20 bg-gray-50 dark:bg-gray-700"
      disabled=""
      placeholder="veronica@example.com"
      type="email"
      value=""
    />
  </div>
</DocumentFragment>
`;
